File: RealmExample/app/src/main/java/com/flatstack/realmexample/base/BaseAdapter.java
package com.flatstack.realmexample.base;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.ViewGroup;

import java.util.List;

import rx.functions.Func1;

public class BaseAdapter<T, VH extends BaseHolder<T>> extends RecyclerView.Adapter<VH> {
    @NonNull private       List<T>              items;
    @NonNull private final Func1<ViewGroup, VH> func0;

    public BaseAdapter(@NonNull List<T> items, @NonNull Func1<ViewGroup, VH> func0) {
        this.items = items;
        this.func0 = func0;
    }

    @Override public VH onCreateViewHolder(ViewGroup viewGroup, int i) {
        return func0.call(viewGroup);
    }

    @Override public void onBindViewHolder(VH vh, int position) {
        vh.setData(items, position);
        vh.draw(items.get(position));
    }

    @Override public int getItemCount() {
        return items.size();
    }

    public List<T> getData() {
        return items;
    }

    public void add(T item) {
        addWithoutNotifying(item);
        notifyItemInserted(items.size() - 1);
    }

    public void addWithoutNotifying(T item) {
        this.items.add(item);
    }
}

File: kotlin-base/app/src/main/java/com/flatstack/android/utils/recycler_view/BaseAdapter.kt
package com.flatstack.android.utils.recycler_view

import android.support.v7.widget.RecyclerView
import android.view.ViewGroup

import rx.functions.Func1

/**
 * Created by ereminilya on 14/12/16.
 */
class BaseAdapter<T, VH : BaseHolder<T>> @JvmOverloads
constructor(private val items: MutableList<T>,
            private val func0: Func1<ViewGroup, VH>,
            private val onItemClickListener: OnItemClickListener<T>? = null)
    : RecyclerView.Adapter<VH>() {

    override fun onCreateViewHolder(viewGroup: ViewGroup, holderType: Int): VH {
        val holder = func0.call(viewGroup)
        holder.itemView.setOnClickListener { v ->
            val position = holder.adapterPosition
            if (position != RecyclerView.NO_POSITION && onItemClickListener != null) {
                onItemClickListener.onItemClick(data[position])
            }
        }
        return holder
    }

    override fun onBindViewHolder(vh: VH, position: Int) = vh.bind(items[position])

    override fun getItemCount() = items.size

    val data: List<T> get() = items

    fun add(item: T) {
        this.items.add(item)
        notifyItemInserted(items.size - 1)
    }
}